feat: Read and write metadata files

loewenheim opened the pull request:
This adds functionality to write and read the metadata structures introduced in #1617 to and from the filesystem.

1. The read happens in `check_expiry`, which gives the cleanup job access to the metadata. All of the old logic is still there—if no metadata file exists, we use the mtime from the FS as before.
2. The write happens in `compute`. The code is pretty ugly because the function now returns `CacheEntry` and therefore can't use `?`; tips for how to make this nicer are very welcome. It should be noted that the shared cache is completely detached from the metadata right now—when we store data in the shared cache, we don't pass the metadata along, and when we read from it, we create fresh metadata as if we had just computed the data ourselves. Propagating metadata into the shared cache is future work.

Open questions:
1. We currently _also_ have the "metadata" .txt files that we only write in debug mode (see the e2e test for an example). We could subsume this into the new metadata model by adding a `"debug"` field and only writing it in debug mode.
2. I put both "time created" and "time last accessed" on `Metadata` for the sake of precision, but it's unclear that we need both, we could do with just time accessed.

Dav1dde commented on crates/symbolicator-service/src/caching/fs.rs:
```diff
@@ -103,39 +104,53 @@ impl Cache {
+                //
+                // * creation time does not exist pre-Linux 4.11
+                // * most filesystems are mounted with noatime
+                let mtime = fs_metadata.modified()?;
```
Only here we have to query the metadata, so I would move the `metadata()` call in here.

Dav1dde replied:
I don't know how hot the function is, but it does save a syscall if you don't unconditionally check it, which may or may not matter.

Dav1dde commented on crates/symbolicator-service/src/caching/fs.rs:
```diff
@@ -193,6 +214,18 @@ impl Cache {
+                        ..metadata.clone()
+                    };
+
+                    write_metadata(path, &new_metadata)?;
```
Having the touch time in the metadata seems quite expensive to upkeep, also it is somewhat of a filesystem specific thing. Some options that come to mind:

1. No longer set the `mtime`, because we already have the `time_accessed`.
2. Keep using the `mtime`, then we don't have to put this into the metadata and the clean-up job still works.
3. Not tracking access time at all. That means we would expire still in use files, would that be terrible?
4. Not tracking access time at all and running the cleanup job from within symbolicator, so we can track the last use time purely in memory.

Not sure how viable 3) and 4) are. 
For 1) I think we should come up with a way to store backend specific metadata, having the access time in the generic metadata seems very filesystem specific.

loewenheim replied:
Right now I'm gravitating towards option 2—don't track `time_accessed` in the metadata, keep using the existing filesystem logic for expiration times.

Dav1dde commented on crates/symbolicator-service/src/caching/fs.rs:
```diff
@@ -201,6 +234,20 @@ impl Cache {
+        catch_not_found(|| {
+            let file = File::open(&md_path)?;
+            let reader = BufReader::new(file);
+            serde_json::from_reader(reader)
```
Maybe (eventually): Just explore what other serialization formats we can use.

loewenheim replied:
MessagePack works well in my local tests, and results in much smaller files.

Dav1dde commented on crates/symbolicator-service/src/caching/fs.rs:
```diff
@@ -330,3 +377,19 @@ where
+/// for a cache file.
+pub(crate) fn metadata_path(path: impl AsRef<Path>) -> PathBuf {
+    let mut out = path.as_ref().to_path_buf();
+    out.set_extension("md");
```
Markdown?^^

Maybe (no clue if `set_extension` allows you to include a dot):
```suggestion
    out.set_extension("metadata.json");
```

Dav1dde commented on crates/symbolicator-service/src/caching/memory.rs:
```diff
@@ -240,20 +251,31 @@ impl<T: CacheItemRequest> Cacher<T> {
                     // Now we have written the data to the tempfile we can mmap it, persisting it later
                     // is fine as it does not move filesystem boundaries there.
-                    let byte_view = ByteView::map_file_ref(temp_file.as_file())?;
+                    let byte_view = match ByteView::map_file_ref(temp_file.as_file()) {
+                        Ok(bv) => bv,
+                        Err(e) => return CacheEntry::from_err(e),
+                    };
```
```suggestion
                    let byte_view = ByteView::map_file_ref(temp_file.as_file()).map_err(CacheEntry::from_err)?;
```

Dav1dde commented on crates/symbolicator-service/src/caching/memory.rs:
```diff
@@ -209,26 +210,36 @@ impl<T: CacheItemRequest> Cacher<T> {
         let shared_cache = self.shared_cache(&request);
         let shared_cache_hit = if let Some(shared_cache) = shared_cache {
-            let temp_fd = tokio::fs::File::from_std(temp_file.reopen()?);
+            // TODO: Here we create fresh metadata upon fetching from the shared cache.
+            // Ideally we would obviously want to share these metadata, but this is left
+            // as future work.
```
Good question, how much we would want to share here. File specific metadata of course, but things like the created timestamp we wouldn't want to necessarily use for on disk cache duration (except maybe for a total max ttl).

Which leads me more to think about again that we might want backend specific metadata and shared metadata. If a backend needs additional metadata to work (Filesystem may need access times), they shouldn't bleed into other backends.

Dav1dde commented on crates/symbolicator-service/src/caching/memory.rs:
```diff
@@ -291,7 +313,16 @@ impl<T: CacheItemRequest> Cacher<T> {
+            };
+
+            // Write metadata if we have them
+            if let Some(metadata) = metadata.as_ref() {
+                if let Err(e) = super::fs::write_metadata(&cache_path, metadata) {
+                    return CacheEntry::from_err(e);
+                };
+            }
```
At this point we return an error, but we still have the cache file persistent on disk. 

I assume on next access this would just be treated as "no metadata available", to be consistent we'd either have to reject caches without metadata or we just swallow the error here or we at least attempt to delete the previously written file.

loewenheim commented:
Changes after feedback:
1. `Metadata` no longer tracks access time. Logic for expiring/cleaning up cache files is as in `master`, i.e. based on FS mtime. This means that the metadata file is effectively immutable (for now).
2. Switched the metadata file format to MessagePack.
3. Switched the metadata file extension to `metadata`. Using something like `metadata.mpk` has the disadvantage that _checking_ the extension would only return the `mpk` part.
4. `write_metadata` deletes the metadata file if writing fails, just to be safe.
5. `cleanup` now deletes metadata files along with cache files.

loewenheim commented:
More changes:
1. Reverted back to JSON for the metadata files
2. Removed logic for file deletion on metadata write failures
